apps/system/js/task_manager.js

```diff
@@ -1,900 +1,519 @@
+          top: 0,
+          behavior: immediately ? 'auto' : 'smooth'
         });
-        return;
-      }
-
-      var app = Service.query('AppWindowManager.getActiveWindow');
-      if (app && !app.isHomescreen) {
-        app.getScreenshot(function onGettingRealtimeScreenshot() {
-          this.show(filter);
-        }.bind(this), 0, 0, 400);
-      } else {
-        this.show(filter);
+        setTimeout(resolve, 200);
       }
-    };
+    });
+  },
 
-    finish();
-  };
+  handleEvent(evt) {
+    var card;
 
-  /**
-   * Handle (synthetic) tap events on the card list
-   *
-   * @memberOf TaskManager.prototype
-   * @param  {DOMEvent} evt The event.
-   */
-  TaskManager.prototype.handleTap = function cs_handleTap(evt) {
-    if (this.element.classList.contains('empty')) {
-      var homescreen = Service.query('getHomescreen', true);
-      this.exitToApp(homescreen);
-      return;
-    }
-
-    var targetNode = evt.target;
-    var card = this.getCardForElement(targetNode);
-    if (!card) {
-      return;
-    }
+    switch (evt.type) {
+      case 'click':
+        if (this.element.classList.contains('empty')) {
+          this.hide(Service.query('getHomescreen', true));
+          return;
+        }
 
-    if ('buttonAction' in targetNode.dataset) {
-      this.cardAction(card, targetNode.dataset.buttonAction);
-      return;
-    }
+        card = this.elementToCardMap.get(evt.target.closest('.card'));
+        if (card) {
+          this.cardAction(card, evt.target.dataset.buttonAction || 'select');
+        }
+        break;
 
-    if (('position' in targetNode.dataset) || card) {
-      this.cardAction(card, 'select');
-      return;
-    }
-  };
+      case 'resize':
+        this.updateLayout();
+        break;
 
-  /**
-   * Gets current sizing information on resize or render.
-   * @memberOf TaskManager.prototype
-   * @param  {DOMEvent} evt The event.
-   */
-  TaskManager.prototype.calculateDimensions =
-    function cv_calculateDimensions(evt) {
-    this.windowWidth = window.innerWidth;
-    this.windowHeight = window.innerHeight;
-    // swipe up gesture should be proportional to current viewport height
-    this.SWIPE_UP_THRESHOLD = this.windowHeight / 4;
-  };
+      case 'scroll':
+        if (this.element.style.overflowX === 'hidden') {
+          // Believe it or not, you will receive scroll events even when
+          // overflow is hidden. We don't care about those, because we're
+          // dragging a card; the user won't see scroll events.
+        } else {
+          this._lastScrollTimestamp = Date.now();
+        }
+        this.updateScrollPosition();
+        break;
 
-  /**
-   * Default event handler
-   * @memberOf TaskManager.prototype
-   * @param  {DOMEvent} evt The event.
-   */
-  TaskManager.prototype.handleEvent = function cv_handleEvent(evt) {
-    var app;
-    switch (evt.type) {
-      case 'click':
-        this.handleTap(evt);
+      case 'card-will-drag':
```

Comment:
we need basic unit testing for the `card-will-drag` / `card-dropped` events (with real dispatch ;))